### PR TITLE
SWP-694 - Use entity uids rather than instances & added additional method of adding cards

### PR DESCRIPTION
**Pull request by adteulade**
- Passing Uid instead of full object so that project that will use this package won't have to call the Stripe API to find object, and allow simpler code on client side.
- Adding backward compatibility with old Source API cards.

### CODE REVIEW

File: tests/Providers/Stripe/Customer/ServiceTest.php [hunk 1]
```diff
@@ -50,9 +51,7 @@ public function testCanAddCard()
             'test'
         );
 
-        $card = $this->service->cards()->find('pm_card_visa');
-
-        $card = $this->service->customers()->addCard($customer, $card);
+        $card = $this->service->customers()->addCard($customer->uid, 'pm_card_visa');
```
Comment (clemblanco) [inline tests/Providers/Stripe/Customer/ServiceTest.php:68]: Why isn't `$this->service` already the CustomerService as you seem to only need that one now?

File: tests/Providers/Stripe/Customer/ServiceTest.php [hunk 2]
```diff
@@ -89,4 +88,24 @@ public function testItCanCreateCustomerWithCardToken()
...
         $this->assertInstanceOf(Customer::class, $customer);
     }
+
+    /**
+     * Test card can be deleted.
+     */
+    public function testCanDeleteCard()
+    {
+        /** @var StripeCustomer $response */
+        $response = StripeCustomer::create([
```
Comment (clemblanco) [inline tests/Providers/Stripe/Customer/ServiceTest.php]: Why not calling `$response` `$customer` if this is of type `StripeCustomer`?

If this is a pure and generic API response, even having an extra variable `$customerId = $response->id;` would make it clearer.

File: tests/Providers/Stripe/Customer/ServiceTest.php [hunk 3]
```diff
@@ -89,4 +88,24 @@ public function testItCanCreateCustomerWithCardToken()
...
+            'email' => 'software@pod-point.com',
+            'description' => 'test',
+        ]);
+
+        $card = $this->service->customers()->addCard($response->id, 'pm_card_visa');
```
Comment (clemblanco) [inline tests/Providers/Stripe/Customer/ServiceTest.php]: Would be nice to have a more fluent service class maybe.
```php
$this->service->customer($customer->id)->addCard('pm_card_visa');
```
or
```php
$this->service->customers()->find($customer->id)->addCard('pm_card_visa');
```
or
```php
$this->service->customers()->forCustomer($customer->id)->addCard('pm_card_visa');
```

File: src/Customer/Service.php [hunk 4]
```diff
@@ -31,29 +31,29 @@ public function find(string $uid): Customer;
...
      *
-     * @param Customer $customer
-     * @param Card $card
+     * @param string $customerUid
+     * @param string $cardUid
      *
      * @return void
      */
-    public function deleteCard(Customer $customer, Card $card): void;
+    public function deleteCard(string $customerUid, string $cardUid): void;
```
Comment (clemblanco) [inline src/Customer/Service.php:49]: Why not using the objects anymore? Objects (even data objects) should always be preferred from primitive types like string.

File: src/Providers/Stripe/Customer/Service.php [hunk 5]
```diff
@@ -64,17 +64,17 @@ public function find(string $uid): Customer
...
+     * @param string $cardUid
      *
      * @return Card
      */
-    public function addCard(Customer $customer, Card $card): Card
+    public function addCard(string $customerUid, string $cardUid): Card
     {
         /** @var PaymentMethod $paymentMethod */
-        $paymentMethod = PaymentMethod::retrieve($card->uid);
+        $paymentMethod = PaymentMethod::retrieve($cardUid);
```
Comment (clemblanco) [inline src/Providers/Stripe/Customer/Service.php]: I wouldn't mind reading `\Stripe\PaymentMethod::retrieve()` instead of `PaymentMethod::retrieve()` for clarity. This way I can see straight away that this is not coming from our package but from Stripe SDK. Would also prevent us having to define aliases when conflicting with out stuff.

File: src/Providers/Stripe/Customer/Service.php [hunk 6]
```diff
@@ -89,29 +89,37 @@ public function addCard(Customer $customer, Card $card): Card
...
-        StripeCustomer::deleteSource($customer->uid, $card->uid);
+        $cardToken = new StripeToken($cardUid);
+
+        if ($cardToken->type === StripeToken::PAYMENT_METHOD) {
+            $paymentMethod = PaymentMethod::retrieve($cardUid);
+
+            $paymentMethod->detach();
+        } else {
+            StripeCustomer::deleteSource($customerUid, $cardUid);
+        }
```
Comment (clemblanco) [inline src/Providers/Stripe/Customer/Service.php]: Are we not returning anything? No feedback whatsoever? Is this synchronous our asynchronous? What if it's failing?

File: src/Providers/Stripe/Customer/Service.php [hunk 7]
```diff
@@ -89,29 +89,37 @@ public function addCard(Customer $customer, Card $card): Card
...
      *
      * @return void
      */
-    public function deleteCard(Customer $customer, Card $card): void
+    public function deleteCard(string $customerUid, string $cardUid): void
     {
-        StripeCustomer::deleteSource($customer->uid, $card->uid);
+        $cardToken = new StripeToken($cardUid);
+
+        if ($cardToken->type === StripeToken::PAYMENT_METHOD) {
```
Comment (clemblanco) [inline src/Providers/Stripe/Customer/Service.php]: could be
```php
if ((new StripeToken($cardUid))->type === StripeToken::PAYMENT_METHOD) {
```
or
```php
if ((new StripeToken($cardUid))->is(StripeToken::PAYMENT_METHOD)) {
```
or
```php
if ((new StripeToken($cardUid))->isPaymentMethod()) {
```

File: src/Card/Service.php [hunk 8]
```diff
@@ -24,11 +24,4 @@ public function find(string $uid): Card;
      * @return Card
      */
     public function create(Token $token = null): Card;
-
-    /**
-     * Deletes a card.
-     *
-     * @param Card $card
-     */
-    public function delete(Card $card): void;
```
Comment (clemblanco) [inline src/Card/Service.php:33]: Moving stuff closer to a Customer as you'll always need to perform this kind of action with a Customer makes sense.

File: src/Customer/Service.php [hunk 9]
```diff
@@ -31,29 +31,29 @@ public function find(string $uid): Customer;
...
 
     /**
      * Retrieves a customers cards.
      *
-     * @param Customer $customer
+     * @param string $uid
```
Comment (clemblanco) [inline src/Customer/Service.php:54]: Event though I don't approve the removal of objects over primitive types, we don't know what kind of identifier this is?

Would it be possible to have it closer to a Customer class rather than a Customer Service class? This way you could call it like:
```php
$foobar->customers()->find($customerUid)->addCard($cardUid);
```
directly on a Customer object as it's an action performed for a Customer.

File: src/Providers/Stripe/Payment/Service.php [hunk 10]
```diff
@@ -51,15 +51,13 @@ public function create(
     ): Payment {
         switch ($token->type) {
             case StripeToken::CUSTOMER:
-                $customer = $this->customers()->find($token->value);
-
-                $cards = $this->customers()->getCards($customer);
+                $cards = $this->customers()->getCards($token->value);
                 $card = $cards[0];
```
Comment (clemblanco) [inline src/Providers/Stripe/Payment/Service.php:56]: `$this->customers()->getCardsForCustomer()` would be more accurate for the current codebase.

Would be nicer to read `getCards()->first()`.
Or `getFirstCard()` if you don't want to bring Collections in.

**Comment by clemblanco** [comment]
Also PR description would have been nice because I didn't really see the connection with "adding setup intent".

File: tests/Providers/Stripe/Customer/ServiceTest.php (hunk 1 above)
Comment (adteulade) [reply tests/Providers/Stripe/Customer/ServiceTest.php:68]: Because if so we'll have to add multiple service via a service provider. So here we decided that the core service which is the PaymentService knows all the other one  and we are able to retrieve it directly from this one.

File: tests/Providers/Stripe/Customer/ServiceTest.php (hunk 2 above)
Comment (adteulade) [reply tests/Providers/Stripe/Customer/ServiceTest.php]: Because we are calling it $response all around the package to differentiate our own Customer object compared to the Stripe Customer object. I'm going to rename it $stripeCustomer if you think it's confusing.

File: tests/Providers/Stripe/Customer/ServiceTest.php (hunk 3 above)
Comment (adteulade) [reply tests/Providers/Stripe/Customer/ServiceTest.php]: I disagree we had a whole discussion about that already and we agreed to do it that way because of the way the Stripe lib is built. Finding a customer before adding the card will add more loading everything is done at once.

File: src/Customer/Service.php (hunk 4 above)
Comment (adteulade) [reply src/Customer/Service.php:49]: Because we'll have to find those object in order to call those method which mean another call to the Stripe API. Whereas Stripe API just needs uids.

File: src/Providers/Stripe/Customer/Service.php (hunk 5 above)
Comment (adteulade) [reply src/Providers/Stripe/Customer/Service.php]: On our package we are using Card and not PaymentMethod. All the naming have already been discussed but happy to have a dev meeting if needed.

File: src/Providers/Stripe/Customer/Service.php (hunk 6 above)
Comment (adteulade) [reply src/Providers/Stripe/Customer/Service.php]: There will be an exception in that case

File: src/Card/Service.php (hunk 8 above)
Comment (adteulade) [reply src/Card/Service.php:33]: FYI You actually won't need a customer with the new Payment Intent Api when you're removing a Payment method which is why it was seperated. But it's just way easier to use that way.

File: src/Customer/Service.php (hunk 9 above)
Comment (adteulade) [reply src/Customer/Service.php:54]: This is the Customer Service so I don't think we need to specify what type of $uid is needed. We are specifying the type of uid needed if there are multiple uid in used in the same method.

**Comment by adteulade** [comment]
Sorry about the description I updated it.

File: src/Providers/Stripe/Payment/Service.php (hunk 10 above)
Comment (adteulade) [reply src/Providers/Stripe/Payment/Service.php:56]: This is a workaround the opencharge web app where we are charging customers directly. We shouldn't do that because we should handle multiple cards in the first place so adding a new method based on a lack of functionality seems irrelevant IMO. But I get your point.

File: src/Providers/Stripe/Customer/Service.php (hunk 7 above)
Comment (adteulade) [reply src/Providers/Stripe/Customer/Service.php]: I'm also going to update it into a switch so that it's easier to read.

File: src/Card.php [hunk 11]
```diff
@@ -48,19 +48,19 @@ class Card
...
-        string $last4,
-        string $brand,
-        string $funding,
-        int $expirationMonth,
-        int $expirationYear
+        string $last4 = null,
+        string $brand = null,
+        string $funding = null,
+        int $expirationMonth = null,
+        int $expirationYear = null
```
Comment (LeeBrooks3) [inline src/Card.php]: as discussed this doesn't feel right...

File: src/Card/Service.php (hunk 8 above)
Comment (LeeBrooks3) [reply src/Card/Service.php:33]: I don't think we should have changed this...

File: src/Customer/Service.php (hunk 9 above)
Comment (LeeBrooks3) [reply src/Customer/Service.php:54]: To add to what @adteulade said @clemblanco we opted for all the "doing" (creating/retrieving/updating/deleting) to be within the services so that the models can remain as plain old php objects.

But also I wouldn't be against this named `customerUid` just to be explicit.

File: src/Providers/Stripe/Card/Service.php [hunk 12]
```diff
@@ -45,8 +45,16 @@ public function find(string $uid): Card
...
-            $response = SetupIntent::retrieve($token->value);
+            switch ($token->type) {
+                case StripeToken::SETUP_INTENT:
+                    /** @var SetupIntent $response */
+                    $response = SetupIntent::retrieve($token->value);
+
+                    break;
+                case StripeToken::TOKEN:
+                default:
+                    return new Card($token->value);
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Card/Service.php]: As discussed, not keen on this as it's not actually creating a card / doing anything.

File: src/Providers/Stripe/Card/Service.php [hunk 13]
```diff
@@ -73,16 +81,4 @@ public function create(Token $token = null): Card
...
-     * Deletes a card using the Stripe SDK.
-     *
-     * @param Card $card
-     */
-    public function delete(Card $card): void
-    {
-        $paymentMethod = PaymentMethod::retrieve($card->uid);
-
-        $paymentMethod->detach();
-    }
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Card/Service.php:87]: As discussed would revert this change.

File: src/Providers/Stripe/Customer/Service.php [hunk 14]
```diff
@@ -33,7 +33,7 @@ public function create(Token $token, string $email, string $description): Custom
                 $params['payment_method'] = $token->value;
 
                 break;
-            case StripeToken::CARD:
+            case StripeToken::TOKEN:
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Customer/Service.php:38]: is there a reason why this was renamed?

File: src/Providers/Stripe/Customer/Service.php [hunk 15]
```diff
@@ -64,54 +64,86 @@ public function find(string $uid): Customer
...
-            $response->card->exp_year
-        );
+        switch ((new StripeToken($cardUid))->type) {
+            case StripeToken::PAYMENT_METHOD:
+                /** @var PaymentMethod $paymentMethod */
+                $paymentMethod = PaymentMethod::retrieve($cardUid);
+
+                $response = $paymentMethod->attach(['customer' => $customerUid]);
+
+                return new Card(
```
Comment (lukasz-kowal) [inline src/Providers/Stripe/Customer/Service.php:89]: I think we are duplicating some code here, personally I would even go for `$response = $response->card`
Not to duplicate card creation.

File: src/Providers/Stripe/Customer/Service.php [hunk 16]
```diff
@@ -64,54 +64,86 @@ public function find(string $uid): Customer
...
-
-        return new Card(
-            $response->id,
-            $response->card->last4,
-            $response->card->brand,
-            $response->card->funding,
-            $response->card->exp_month,
-            $response->card->exp_year
-        );
+        switch ((new StripeToken($cardUid))->type) {
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Customer/Service.php]: would change this to:
```
$token = new StripeToken($cardUid);

switch ($token->type) {
```

File: src/Providers/Stripe/Customer/Service.php (hunk 14 above)
Comment (adteulade) [reply src/Providers/Stripe/Customer/Service.php:38]: Yeah I wasn't using the right token type the CARD token is never going to be used in that case.

File: src/Providers/Stripe/Customer/Service.php (hunk 15 above)
Comment (adteulade) [reply src/Providers/Stripe/Customer/Service.php:89]: As discussed the id is not part of the card object. Also I tried to modify the response object and Stripe lib won't let you do it so the solution would be to create an array but it will add more code for the same result.

File: src/Providers/Stripe/Customer/Service.php (hunk 6 above)
Comment (clemblanco) [reply src/Providers/Stripe/Customer/Service.php]: A PHP Exception? We would need to add `@throws` in the docblock then.

File: tests/Providers/Stripe/Customer/ServiceTest.php (hunk 3 above)
Comment (clemblanco) [reply tests/Providers/Stripe/Customer/ServiceTest.php]: I get your point about avoiding multiple API calls but this is different from having this using fluent functions.

Comment (adteulade) [reply tests/Providers/Stripe/Customer/ServiceTest.php]: As @LeeBrooks3 explained:

> we opted for all the "doing" (creating/retrieving/updating/deleting) to be within the services so that the models can remain as plain old php objects.

File: src/Providers/Stripe/Customer/Service.php (hunk 6 above)
Comment (adteulade) [reply src/Providers/Stripe/Customer/Service.php]: The thing is the Stripe\Error\InvalidRequest could happened everywhere we are calling Stripe API. So I thought it wasn't necessary to specify all those `@throws` on every method that is using the Stripe API. What do you think? Happy to add it everywhere if needed.

Edit: adding missing @throws.

**Review by lukasz-kowal — APPROVED** [review]
Looks good to me, except duplicate delete method we spoke about.

File: src/Card/Service.php (hunk 8 above)
Comment (clemblanco) [reply src/Card/Service.php:33]: Why? Backward compatibility issues?

Comment (adteulade) [reply src/Card/Service.php:33]: So there are two delete methods one for the Source API cards and one for the new Payment Method API. I regrouped both in one but wasn't a good idea in the end.

File: src/Providers/Stripe/Card/Service.php (hunk 13 above)
Comment (clemblanco) [reply src/Providers/Stripe/Card/Service.php:87]: Why?

Comment (adteulade) [reply src/Providers/Stripe/Card/Service.php:87]: > So there are two delete methods one for the Source API cards and one for the new Payment Method API. I regrouped both in one but wasn't a good idea in the end.

File: src/Customer/Service.php (hunk 9 above)
Comment (clemblanco) [reply src/Customer/Service.php:54]: > we opted for all the "doing" (creating/retrieving/updating/deleting) to be within the services

Could still happen with fluent methods.

**Review by clemblanco — APPROVED** [review]

File: src/Customer/Service.php (hunk 9 above)
Comment (adteulade) [reply src/Customer/Service.php:54]: @clemblanco First sorry for being annoying with that but:
Based on:

> ... so that the models can remain as plain old php objects.

If we are using:

> $foobar->customers()->find($customerUid)->addCard($cardUid);

When calling `find()` we'll get a Customer object which mean having `addCards()` using the Customer model will imply that models contains logic code instead of being plain old php objects.

**Review by LeeBrooks3 — APPROVED** [review]
